src/mscorlib/src/System/Reflection/Emit/DynamicILGenerator.cs

```diff
@@ -430,24 +433,14 @@ public override void BeginCatchBlock(Type exceptionType)
 
                 // if this is a catch block the exception will be pushed on the stack and we need to update the stack info
                 UpdateStackSize(OpCodes.Nop, 1);
-            }
 
-            current.MarkCatchAddr(ILOffset, exceptionType);
+                current.MarkCatchAddr(ILOffset, exceptionType);
```

Comment:
I meant something like this. Same flow as the base-class, the only differences are the special handling of tokens for dynamic methods, and the UpdateStackSize bugfix.

```
   if (current.GetCurrentState() == __ExceptionInfo.State_Filter)
   {
       ...existing code...
   }
   else
   {
       ...existing code...

       // this is relying on too much implementation details of the base and so it's highly breaking 
       // Need to have a more integreted story for exceptions 
       current.m_filterAddr[current.m_currentCatch - 1] = GetTokenFor(rtType); 
   }

   current.MarkCatchAddr(ILOffset, exceptionType);

   // if this is a catch block the exception will be pushed on the stack and we need to update the stack info
   UpdateStackSize(OpCodes.Nop, 1);
```